cirq/ops/swap_gates.py

```diff
@@ -85,16 +85,32 @@ def _trace_distance_bound_(self) -> Optional[float]:
             return None
         return abs(np.sin(self._exponent * 0.5 * np.pi))
 
+    def _has_stabilizer_effect_(self) -> Optional[bool]:
+        if self._is_parameterized_():
+            return None
+        return self.exponent % 1 == 0
+
     def _act_on_(self, args):
         from cirq import ops, sim, protocols
 
-        if isinstance(args, sim.ActOnStabilizerCHFormArgs) and self._exponent % 2 == 1:
-            args.state.omega *= 1j ** (2 * self.global_shift * self._exponent)
-            protocols.act_on(ops.CNOT, args)
-            args.axes = args.axes[::-1]
-            protocols.act_on(ops.CNOT, args)
-            args.axes = args.axes[::-1]
-            protocols.act_on(ops.CNOT, args)
+        if isinstance(args, (sim.ActOnStabilizerCHFormArgs, sim.ActOnCliffordTableauArgs)):
+            if not protocols.has_stabilizer_effect(self):
+                return NotImplemented
+            if isinstance(args, sim.ActOnStabilizerCHFormArgs):
+                args.state.omega *= 1j ** (2 * self.global_shift * self._exponent)
+
+            if self._exponent % 2 == 1:
+                protocols.act_on(ops.CNOT, args)
+                args.axes = args.axes[::-1]
+                protocols.act_on(ops.CNOT, args)
+                args.axes = args.axes[::-1]
+                protocols.act_on(ops.CNOT, args)
+            else:
+                # Any fractional exponent SWAP gate should have been rejected
+                # already. Any even exponent does not change anything except
+                # the global phase above.
+                assert self._exponent % 2 == 0
```

Comment:
I'd drop the assert. Especially after the has stabilizer effect method is inlined.

Reply:
Done